File: software/lua/modules/osg/land/TerrainNode.lua
local Class = require("classBuilder"){name="TerrainNode",bases="base.Object"};

require "osg"
require "sgtLand"

local HORIZON_SIZE = 256

function Class:initialize(options)	
	self:info("Creating TerrainNode.")
	self._root = osg.Group()
	self._transform = osg.MatrixTransform()
	self._root:addChild(self._transform)
	
	self._ss = self._root:getOrCreateStateSet()
	
	-- attach to parent if provide:
	if options and options.parent then
		options.parent:addChild(self._root)
	end
	
	-- Create the deformation object:
	local Deformation = require "osg.land.Deformation"
	self._deform = Deformation{node=self}
	
	self._splitFactor = options and options.splitFactor or 2.0;
	self._splitDist = 1.1; -- the viewer distance at which the quad is subdivided.
	self._splitInvisibleQuads = false;
    self._horizonCulling = true;
	self._maxLevel = 7;
	self._horizon = {};
	self._frustumPlanes = osg.PlaneList() 
	
	--self._deformedCameraPos = osg.Vec3d();
	--self._localCameraPos = osg.Vec3d();
	
	-- create the shared tile geode:
	self:info("Creating Tile geode.")
	self:createTileGeode()
	
	-- create the quad tree:
	self:info("Creating root TerrainQuad.")
	local TerrainQuad = require "osg.land.TerrainQuad"
	self._quad = TerrainQuad{node=self}
	self._transform:addChild(self._quad:getRoot())
	
	self:_root:setCullCB(function(node,nv)
		-- compute the local to world mat:
		--local ltow = osg.computeLocalToWorld(nv:getNodePath())
		
		local cv = nv:dynCast("osgUtil::CullVisitor")
		self:check(cv,"Invalid cull visitor.")
		
		local mv = cv:getModelViewMatrix()
		local proj = cv:getProjectionMatrix();
		local cullSet = cv:getCurrentCullingSet();
		local viewport = cv:getViewport();
		self:check(viewport,"Invalid viewport object.")
		
		-- retrieve the frustum planes:
		-- local frustum = cullSet:getFrustum();
		-- for i = 0,5 do
			-- self._frustumPlanes[i+1] = frustum:
		-- end
		sgt.getFrustumPlanes(sgt.mat4d(proj),self._frustumPlanes)
		
		local res, fovy, aspect, znear, zfar = proj:getPerspective(0.0,0.0,0.0,0.0);
		self:check(res,"Could not get perspective settings from projection matrix ",proj)
		
		-- compute the fovx angle:
		-- The computed relation is:
		-- fovx = 2 * atan( aspect * tan(fovy/2))
		local fovx = 2.0 * math.atan(aspect* math.tan(math.rad(fovy/2.0))); -- in radians.
		
		self._projection = proj;
		--local mvp = mv*proj;
		
		local mv_inv = osg.Matrixd.inverse(mv);
		self._deformedCameraPos = mv_inv:getTrans();

		self._localCameraPos = self._deform:deformedToLocal(self._deformedCameraPos);

		local m = deform:localToDeformedDifferential(self._localCameraPos, true);
		self._distFactor = max(m:getRow3(0):length(), m:getRow3(1):length());

		self._splitDist = self._splitFactor * viewport:width() / 1024.0 * tan(math.rad(40.0)) / tan(fovx / 2.0);
		if (splitDist < 1.1 or not stg.isFinite(splitDist)) then
			splitDist = 1.1;
		end

		-- initializes data structures for horizon occlusion culling
		if (self._horizonCulling and self._localCameraPos:z() <= self._quad:getZmax()) then
			local deformedDir = mv_inv:preMult(osg.ZAXIS);
			local localDir = (self._deform:deformedToLocal(deformedDir) - self._localCameraPos).xy()
			localDir:normalize();
			
			self._localCameraDir = sgt.mat2f(localDir:getY(), -localDir:getX(), -localDir:getX(), -localDir:getY());
			for i = 1,HORIZON_SIZE do
				self._horizon[i] = -sgt.INFINITY;
			end
		end

		self._quad:update();		
	end)
end

function Class:getVisibility(box)
	return self._deform:getVisibility(box)
end

function Class:getDeformedFrustumPlanes()
	self:check(self._frustumPlanes:size() == 6, "Invalid frustum planes.")
	return self._frustumPlanes
end

function Class:getDeformedCamera()
	self:check(self._deformedCameraPos,"Invalid deformedCameraPos vector")
	return self._deformedCameraPos
end

function Class:getSplitDistance()
    self:check(self._splitDist > 1.0,"Invalid splidDist value: ",self._splitDist);
    return self._splitDist;
end

function Class:getCameraToScreen()
	self:check(self._projection,"Invalid projection matrix")
	return self._projection
end

function Class:getLocalCamera()
	self:check(self._localCameraPos,"Invalid localCameraPos vector")
	return self._localCameraPos
end

function Class:isOccluded(box)
    if (not self._horizonCulling or self._localCameraPos:z() > self._quad:getZmax()) then
        return false;
    end
	
    local corners = {};
    local o = self._localCameraPos:xy();
    corners[0] = self._localCameraDir * (sgt.vec2d(box:getXmin(), box:getYmin()) - o);
    corners[1] = self._localCameraDir * (sgt.vec2d(box:getXmin(), box:getYmax()) - o);
    corners[2] = self._localCameraDir * (sgt.vec2d(box:getXmax(), box:getYmin()) - o);
    corners[3] = self._localCameraDir * (sgt.vec2d(box.getXmax(), box:getYmax()) - o);
    if (corners[0]:getY() <= 0.0 or corners[1]:getY() <= 0.0 or corners[2]:getY() <= 0.0 or corners[3]:getY() <= 0.0) then
        return false;
    end
	
    local dz = float(box.getZmax() - self._localCameraPos:z());
    corners[0] = sgt.vec2f(corners[0]:getX(), dz) / corners[0]:getY();
    corners[1] = sgt.vec2f(corners[1]:getX(), dz) / corners[1]:getY();
    corners[2] = sgt.vec2f(corners[2]:getX(), dz) / corners[2]:getY();
    corners[3] = sgt.vec2f(corners[3]:getX(), dz) / corners[3]:getY();
    local xmin = math.min(corners[0]:getX(), corners[1]:getX(), corners[2]:getX(), corners[3]:getX()) * 0.33f + 0.5f;
    local xmax = math.max(corners[0]:getX(), corners[1]:getX(), corners[2]:getX(), corners[3]:getX()) * 0.33f + 0.5f;
    local zmax = math.max(corners[0]:getY(), corners[1]:getY(), corners[2]:getY(), corners[3]:getY());
    local imin = math.max(math.floor(xmin * HORIZON_SIZE), 0);
    local imax = math.min(math.ceil(xmax * HORIZON_SIZE), HORIZON_SIZE - 1);
    for i = imin,imax do
        if (zmax > self._horizon[i+1]) then
            return false;
        end
    end
	
    return imax >= imin;
end

function Class:getLocalToCamera()
	self:no_impl()
end

function Class:getLocalToWorld()
	self:no_impl()
end

function Class:createTileGeode()
	self._tileGeode = osg.Geode()
	
	local geom = osg.Geometry()
	geom:setUseVertexBufferObjects(true)
	self._tileGeode:addDrwable(geom)
	
	local vertices = osg.Vec3Array()
	vertices:push_back(osg.Vec3f(0.0,0.0,0.0))
	vertices:push_back(osg.Vec3f(1.0,0.0,0.0))
	vertices:push_back(osg.Vec3f(0.0,1.0,0.0))
	vertices:push_back(osg.Vec3f(1.0,1.0,0.0))
	
	geom:setVertexArray(vertices)
	geom:addPrimitiveSet(osg.DrawArrays(osg.PrimitiveSet.TRIANGLE_STRIP,0,4));
end

function Class:setupProgram()
	local shman = require "osg.ShaderManager"
	local prog = shman:loadProgram{"terrain_shader_minimal"}
	self._ss:setAttributeAndModes(prog)	
end

function Class:getTileGeode()
	self:check(self._tileGeode,"Invalid tile geode.")
	return self._tileGeode
end

function Class:getRoot()
	return self._root
end

return Class

File: sources/sgtLand/src/terrain/TerrainNode.cpp
#include "terrain/TerrainNode.h"
#include "osgUtil/CullVisitor"

#define HORIZON_SIZE 256

namespace sgt {

class TerrainCullCB : public osg::NodeCallback {
public:
	virtual void operator() (osg::Node *node, osg::NodeVisitor *nv) {
		TerrainNode* t = dynamic_cast<TerrainNode*>(node);
		CHECK(t,"Invalid terrain node object");
		
		osgUtil::CullVisitor* cv = dynamic_cast<osgUtil::CullVisitor*>(nv);
		CHECK(cv,"Invalid cull visitor object");

		osg::Matrixd ltow = osg::computeLocalToWorld(nv->getNodePath());
		osg::Matrixd viewMatrix = osg::Matrixd::inverse(ltow)*(*cv->getModelViewMatrix());
		
		t->update(ltow,viewMatrix,*cv->getProjectionMatrix(),cv->getViewport());
		
		traverse(node,nv);
	}
};

TerrainNode::TerrainNode(Deformation* deform, osg::Geode* tile, double half_size, double zmin, double zmax, float splitFactor, int maxLevel) 
: deform(deform), tileGeode(tile), splitFactor(splitFactor), maxLevel(maxLevel) {
	THROW_IF(!deform,"Invalid deformation for terrain node.");
	THROW_IF(!tile,"Invalid tile geode for terrain node.");
	root = new TerrainQuad(this, NULL, 0, 0, -half_size, -half_size, 2.0 * half_size, zmin, zmax);
	splitInvisibleQuads = false;
	horizonCulling = true;
	splitDist = 1.1f;
	horizon = new float[HORIZON_SIZE];
	
	// Assign a cull callback to this node:
	setCullCallback(new TerrainCullCB);
	setCullingActive(false); // no culling for this node.

	// Create the needed uniforms:
	osg::StateSet* ss = getOrCreateStateSet();
	blendingU = new osg::Uniform(osg::Uniform::FLOAT_VEC2,"deformation.blending");
	ss->addUniform(blendingU);

	localToScreenU = new osg::Uniform(osg::Uniform::FLOAT_MAT4,"deformation.localToScreen");
	ss->addUniform(localToScreenU);
}

TerrainNode::~TerrainNode() {
	delete [] horizon;
}

vec3d TerrainNode::getDeformedCamera() const
{
	return deformedCameraPos;
}

const osg::Polytope::PlaneList& TerrainNode::getDeformedFrustumPlanes() const
{
	return frustumPlanes;
}

vec3d TerrainNode::getLocalCamera() const
{
	return localCameraPos;
}

float TerrainNode::getCameraDist(const box3d &localBox) const
{
	return (float) max(abs(localCameraPos.z - localBox.zmax) / distFactor,
		max(min(abs(localCameraPos.x - localBox.xmin), abs(localCameraPos.x - localBox.xmax)),
		min(abs(localCameraPos.y - localBox.ymin), abs(localCameraPos.y - localBox.ymax))));
}

LandManager::Visibility TerrainNode::getVisibility(const box3d &localBox) const
{
	return deform->getVisibility(this, localBox);
}

float TerrainNode::getSplitDistance() const
{
	CHECK_RET(isFinite(splitDist),0.0f,"Infinite splitDist.");
	CHECK_RET(splitDist > 1.0f,0.0f,"SplitDist too small: "<<splitDist);
	return splitDist;
}

float TerrainNode::getDistFactor() const
{
	return distFactor;
}

void TerrainNode::update(const mat4d& ltow, const mat4d& viewMatrix, const mat4d& proj, osg::Viewport* vp)
{
	CHECK(vp,"Invalid viewport argument.");

	mat4d modelView = viewMatrix*ltow;

	localToCamera = modelView;
	cameraToScreen = proj;
	localToWorld = ltow;
	cameraWorldPos = viewMatrix.inverse()*vec3d.ZERO;

	mat4d mvp = proj*modelView;
	mat4d modelViewInv = modelView.inverse();

	deformedCameraPos = modelViewInv*vec3d.ZERO;
	LandManager::getFrustumPlanes(mvp, frustumPlanes);
	localCameraPos = deform->deformedToLocal(deformedCameraPos);

	mat4d m = deform->localToDeformedDifferential(localCameraPos, true);
	distFactor = max(vec3d(m[0][0], m[1][0], m[2][0]).length(), vec3d(m[0][1], m[1][1], m[2][1]).length());

	osg::Vec3d left = frustumPlanes[0].getNormal();
	osg::Vec3d right = frustumPlanes[1].getNormal();
	left.normalize();
	right.normalize();

	float fov = (float) safe_acos(-left*right);
	splitDist = splitFactor * vp->width() / 1024.0f * tan(osg::DegreesToRadians(40.0f)) / tan(fov / 2.0f);
	if (splitDist < 1.1f || !(isFinite(splitDist))) {
		splitDist = 1.1f;
	}

	// initializes data structures for horizon occlusion culling
	if (horizonCulling && localCameraPos.z <= root->zmax) {
		vec3d deformedDir = modelViewInv * vec3d.UNIT_Z;
		vec2d localDir = (deform->deformedToLocal(deformedDir) - localCameraPos).xy().normalize();
		localCameraDir = mat2f(localDir.y, -localDir.x, -localDir.x, -localDir.y);
		for (int i = 0; i < HORIZON_SIZE; ++i) {
			horizon[i] = -INFINITY;
		}
	}

	// update the default uniforms:
	deform->setUniforms(this);
	
	root->update();
}

bool TerrainNode::addOccluder(const box3d &occluder)
{
	if (!horizonCulling || localCameraPos.z > root->zmax) {
		return false;
	}
	vec2f corners[4];
	vec2d o = vec2d(localCameraPos.x,localCameraPos.y);
	corners[0] = localCameraDir * (vec2d(occluder.xmin, occluder.ymin) - o).cast<float>();
	corners[1] = localCameraDir * (vec2d(occluder.xmin, occluder.ymax) - o).cast<float>();
	corners[2] = localCameraDir * (vec2d(occluder.xmax, occluder.ymin) - o).cast<float>();
	corners[3] = localCameraDir * (vec2d(occluder.xmax, occluder.ymax) - o).cast<float>();
	if (corners[0].y <= 0.0f || corners[1].y <= 0.0f || corners[2].y <= 0.0f || corners[3].y <= 0.0f) {
		// skips bounding boxes that are not fully behind the "near plane"
		// of the reference frame used for horizon occlusion culling
		return false;
	}
	float dzmin = float(occluder.zmin - localCameraPos.z);
	float dzmax = float(occluder.zmax - localCameraPos.z);
	vec3f bounds[4];
	bounds[0] = vec3f(corners[0].x, dzmin, dzmax) / corners[0].y;
	bounds[1] = vec3f(corners[1].x, dzmin, dzmax) / corners[1].y;
	bounds[2] = vec3f(corners[2].x, dzmin, dzmax) / corners[2].y;
	bounds[3] = vec3f(corners[3].x, dzmin, dzmax) / corners[3].y;
	float xmin = min(min(bounds[0].x, bounds[1].x), min(bounds[2].x, bounds[3].x)) * 0.33f + 0.5f;
	float xmax = max(max(bounds[0].x, bounds[1].x), max(bounds[2].x, bounds[3].x)) * 0.33f + 0.5f;
	float zmin = min(min(bounds[0].y, bounds[1].y), min(bounds[2].y, bounds[3].y));
	float zmax = max(max(bounds[0].z, bounds[1].z), max(bounds[2].z, bounds[3].z));

	int imin = max(int(floor(xmin * HORIZON_SIZE)), 0);
	int imax = min(int(ceil(xmax * HORIZON_SIZE)), HORIZON_SIZE - 1);

	// first checks if the bounding box projection is below the current horizon line
	bool occluded = imax >= imin;
	for (int i = imin; i <= imax; ++i) {
		if (zmax > horizon[i]) {
			occluded = false;
			break;
		}
	}
	if (!occluded) {
		// if it is not, updates the horizon line with the projection of this bounding box
		imin = max(int(ceil(xmin * HORIZON_SIZE)), 0);
		imax = min(int(floor(xmax * HORIZON_SIZE)), HORIZON_SIZE - 1);
		for (int i = imin; i <= imax; ++i) {
			horizon[i] = max(horizon[i], zmin);
		}
	}
	return occluded;
}

bool TerrainNode::isOccluded(const box3d &box)
{
	if (!horizonCulling || localCameraPos.z > root->zmax) {
		return false;
	}
	vec2f corners[4];
	vec2d o = vec2d(localCameraPos.x,localCameraPos.y);
	corners[0] = localCameraDir * (vec2d(box.xmin, box.ymin) - o).cast<float>();
	corners[1] = localCameraDir * (vec2d(box.xmin, box.ymax) - o).cast<float>();
	corners[2] = localCameraDir * (vec2d(box.xmax, box.ymin) - o).cast<float>();
	corners[3] = localCameraDir * (vec2d(box.xmax, box.ymax) - o).cast<float>();
	if (corners[0].y <= 0.0f || corners[1].y <= 0.0f || corners[2].y <= 0.0f || corners[3].y <= 0.0f) {
		return false;
	}
	float dz = float(box.zmax - localCameraPos.z);
	corners[0] = vec2f(corners[0].x, dz) / corners[0].y;
	corners[1] = vec2f(corners[1].x, dz) / corners[1].y;
	corners[2] = vec2f(corners[2].x, dz) / corners[2].y;
	corners[3] = vec2f(corners[3].x, dz) / corners[3].y;
	float xmin = min(min(corners[0].x, corners[1].x), min(corners[2].x, corners[3].x)) * 0.33f + 0.5f;
	float xmax = max(max(corners[0].x, corners[1].x), max(corners[2].x, corners[3].x)) * 0.33f + 0.5f;
	float zmax = max(max(corners[0].y, corners[1].y), max(corners[2].y, corners[3].y));
	int imin = max(int(floor(xmin * HORIZON_SIZE)), 0);
	int imax = min(int(ceil(xmax * HORIZON_SIZE)), HORIZON_SIZE - 1);
	for (int i = imin; i <= imax; ++i) {
		if (zmax > horizon[i]) {
			return false;
		}
	}
	return imax >= imin;
}

void TerrainNode::traverse(osg::NodeVisitor& nv)
{
	// traverse the quad tree.
	root->accept(nv);
}

osg::BoundingSphere TerrainNode::computeBound() const
{
	// Compute the bounding sphere:
	// we should take into account the deformation here.
	osg::BoundingSphere bsphere;
	//bsphere.radius() = max(root->l*sqrt(2.0),max(abs(root->zmax),abs(root->zmin)));
	trWARN("TerrainNode","computeBound() is not implemented.");
	return bsphere;
}

};

